src/Elements/Spatial/Grid2d.cs

```diff
@@ -134,6 +138,11 @@ public Grid2d(IList<Polygon> boundaries, Transform transform = null)
             toGrid.Invert();
 
             var transformedBoundaries = toGrid.OfPolygons(boundaries);
+            var allVerticesZ = transformedBoundaries.SelectMany(b => b.Vertices).Select(v => v.Z);
```

Comment:
I've talked a bit with @wynged about not wanting Linq if we can avoid it, but especially in geometry code. There are performance penalties to Linq that are subtle: https://stackoverflow.com/questions/14893924/for-vs-linq-performance-vs-future. I am happy to be proved wrong about this. Can we do this as a set of nested for loops with an early out?

Reply:
done